### PR TITLE
Twig 2 5548 fixes

### PR DESCRIPTION
**Pull request by andkovt**

### CODE REVIEW

File: src/TwigJs/Compiler/Expression/NullCoalesceExpression.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+use TwigJs\JsCompiler;
+use TwigJs\TypeCompilerInterface;
+
+class NullCoalesceExpression implements TypeCompilerInterface
+{
+
+    public function getType()
+    {
+        return 'Twig\Node\Expression\NullCoalesceExpression';
```
Comment (oleg-andreyev) [inline src/TwigJs/Compiler/Expression/NullCoalesceExpression.php]: `::class`

File: src/TwigJs/Compiler/Expression/NullCoalesceExpression.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    public function compile(JsCompiler $compiler, \Twig_NodeInterface $node)
+    {
+        if (!$node instanceof \Twig_Node_Expression_Conditional) {
+            throw new \RuntimeException(
+                sprintf(
+                    '$node must be an instanceof of \Expression_Conditional, but got "%s".',
```
Comment (oleg-andreyev) [inline src/TwigJs/Compiler/Expression/NullCoalesceExpression.php]: wrong class

File: src/TwigJs/Compiler/Expression/NullCoalesceExpression.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+{
+
+    public function getType()
+    {
+        return 'Twig\Node\Expression\NullCoalesceExpression';
+    }
+
+    public function compile(JsCompiler $compiler, \Twig_NodeInterface $node)
+    {
+        if (!$node instanceof \Twig_Node_Expression_Conditional) {
```
Comment (oleg-andreyev) [inline src/TwigJs/Compiler/Expression/NullCoalesceExpression.php]: deprecaed class

File: src/TwigJs/Compiler/Expression/NullCoalesceExpression.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+class NullCoalesceExpression implements TypeCompilerInterface
+{
+
+    public function getType()
+    {
+        return 'Twig\Node\Expression\NullCoalesceExpression';
+    }
+
+    public function compile(JsCompiler $compiler, \Twig_NodeInterface $node)
```
Comment (oleg-andreyev) [inline src/TwigJs/Compiler/Expression/NullCoalesceExpression.php]: deprecated interface

File: src/TwigJs/Compiler/ForCompiler.php [hunk 5]
```diff
@@ -70,51 +70,43 @@ public function compile(JsCompiler $compiler, Node $node)
...
         if ($node->hasNode('else') && null !== $node->getNode('else')) {
             $compiler->write("var $iteratedName = false;\n");
         }
 
-        if ($node->getAttribute('with_loop')) {
-            $compiler
-                ->write("var $loopName = {\n")
-                ->indent()
-            ;
+        $compiler
```
Comment (oleg-andreyev) [inline src/TwigJs/Compiler/ForCompiler.php]: it's better to build an array and then just `json_encode` + pretty

File: src/TwigJs/Compiler/IncludeCompiler.php [hunk 6]
```diff
@@ -53,21 +53,27 @@ public function compile(JsCompiler $compiler, Node $node)
...
 //         }
 
-        $compiler->isTemplateName = true;
+        $compiler->setTemplateName(true);
         if ($node->getNode('expr') instanceof ConstantExpression) {
             $compiler
                 ->write("(new ")
+                ->write('twig.templates[')
                 ->subcompile($node->getNode('expr'))
+                ->raw(".replace('.twig', '')")
```
Comment (oleg-andreyev) [inline src/TwigJs/Compiler/IncludeCompiler.php]: looks like it's better to leave `.twig` in the template name

File: src/TwigJs/JsCompiler.php [hunk 7]
```diff
@@ -215,6 +215,7 @@ public function __construct(Environment $env)
             ConstantExpression::class => new ConstantCompiler(),
             GetAttrExpression::class => new GetAttrCompiler(),
             MethodCallExpression::class => new MethodCallCompiler(),
+            'Twig\Node\Expression\NullCoalesceExpression' => new Compiler\Expression\NullCoalesceExpression(),
```
Comment (oleg-andreyev) [inline src/TwigJs/JsCompiler.php]: `::class`

File: src/TwigJs/JsCompiler.php [hunk 8]
```diff
@@ -255,12 +256,13 @@ public function __construct(Environment $env)
...
-            'none' => new NoneCompiler(),
-            'null' => new NullCompiler(),
-            'odd' => new OddCompiler(),
-            'sameas' => new SameAsCompiler(),
+            'empty'       => new EmptyCompiler(),
+            'even'        => new EvenCompiler(),
+            'none'        => new NoneCompiler(),
+            'null'        => new NullCompiler(),
+            'odd'         => new OddCompiler(),
+            'sameas'      => new SameAsCompiler(),
```
Comment (oleg-andreyev) [inline src/TwigJs/JsCompiler.php]: `sameas` should be removed. as it's deprecated in 1.x and removed in 2.x

**Review by oleg-andreyev — CHANGES_REQUESTED** [review]

File: src/TwigJs/Compiler/Expression/NullCoalesceExpression.php (hunk 1 above)
Comment (andkovt) [reply src/TwigJs/Compiler/Expression/NullCoalesceExpression.php]: Done

File: src/TwigJs/Compiler/Expression/NullCoalesceExpression.php (hunk 2 above)
Comment (andkovt) [reply src/TwigJs/Compiler/Expression/NullCoalesceExpression.php]: Fixed

File: src/TwigJs/Compiler/Expression/NullCoalesceExpression.php (hunk 3 above)
Comment (andkovt) [reply src/TwigJs/Compiler/Expression/NullCoalesceExpression.php]: Fixed

File: src/TwigJs/Compiler/Expression/NullCoalesceExpression.php (hunk 4 above)
Comment (andkovt) [reply src/TwigJs/Compiler/Expression/NullCoalesceExpression.php]: Fixed

File: src/TwigJs/Compiler/ForCompiler.php (hunk 5 above)
Comment (andkovt) [reply src/TwigJs/Compiler/ForCompiler.php]: Using json_encode

File: src/TwigJs/JsCompiler.php (hunk 7 above)
Comment (andkovt) [reply src/TwigJs/JsCompiler.php]: Fixed

File: src/TwigJs/JsCompiler.php (hunk 8 above)
Comment (andkovt) [reply src/TwigJs/JsCompiler.php]: Removed

**Comment by oleg-andreyev** [comment]
@andkovt tests are failing.

**Pull request merged**